WeakAuras/WeakAuras.lua

```diff
@@ -3447,52 +3447,84 @@ function WeakAuras.ValidateUniqueDataIds(silent)
+  -- 2. for each aura with data.controlledChildren, check that each controlledChild is unique in that table
+  --    and that the corresponding child aura has the correct parent in data.parent. Delete any entries which don't work
+  local parents = {}
+  local children = {}
+  local childHasParent = {}
   for id, data in pairs(db.displays) do
-    if(data.parent) then
-      if(data.controlledChildren) then
-        if not(silent) then
-          print("|cFF8800FFWeakAuras|r detected desynchronization in saved variables:", id, "has both child and parent");
+    if data.parent then
+      if data.controlledChildren then
+        if not silent then
+          print("|cFF8800FFWeakAuras|r detected corruption in saved variables:", id, "has both child and parent.")
         end
         -- A display cannot have both children and a parent
-        data.parent = nil;
-      elseif(db.displays[data.parent] and db.displays[data.parent].controlledChildren) then
-        childToParent[id] = data.parent;
-        parentToChild[data.parent] = parentToChild[data.parent] or {};
-        parentToChild[data.parent][id] = true;
-      else
+        data.parent = nil
+      elseif not db.displays[data.parent] then
         if not(silent) then
-          print("|cFF8800FFWeakAuras|r detected desynchronization in saved variables:", id, "has a nonexistent parent");
+          print("|cFF8800FFWeakAuras|r detected corruption in saved variables:", id, "has a nonexistent parent")
         end
-        data.parent = nil;
+        data.parent = nil
+      elseif not db.displays[data.parent].controlledChildren then
+        if not silent then
+          prettyPrint("detected corruption in saved variables: "..id.." thinks "..data.parent..
+                      " controls it, but "..data.parent.." is not a group.")
+        end
+        data.parent = nil
+      else
+        children[id] = data
       end
+    elseif data.controlledChildren then
+      parents[id] = data
     end
   end
 
-  for id, data in pairs(db.displays) do
-    if(data.controlledChildren) then
-      for index, childId in pairs(data.controlledChildren) do
-        if not(childToParent[childId] and childToParent[childId] == id) then
-          if not(silent) then
-            print("|cFF8800FFWeakAuras|r detected desynchronization in saved variables:", id, "thinks it controls", childId, "but does not");
-          end
-          tremove(data.controlledChildren, index);
+  for id, data in pairs(parents) do
+    local groupChildren = {}
+    local childrenToRemove = {}
+    for index, childID in ipairs(data.controlledChildren) do
+      local child = children[childID]
+      if not child then
+        if not silent then
+          prettyPrint("detected corruption in saved variables: "..id.." thinks it controls "..childID.." which doesn't exist.")
         end
+        childrenToRemove[index] = true
+      elseif child.parent ~= id then
```

Comment:
Shouldn't you do `childrenToRemove[index] = true` for this case ?

Reply:
Yes, that is correct. Thanks for noticing :D